test/e2e_node/container_lifecycle_pod_construction.go

```diff
@@ -124,28 +124,34 @@ func (o containerOutputList) String() string {
 
 // RunTogether returns an error the lhs and rhs run together
 func (o containerOutputList) RunTogether(lhs, rhs string) error {
-	lhsStart := o.findIndex(lhs, "Started", 0)
-	rhsStart := o.findIndex(rhs, "Started", lhsStart+1)
-
-	lhsFinish := o.findIndex(lhs, "Exiting", lhsStart+1)
-	rhsFinish := o.findIndex(rhs, "Exiting", rhsStart+1)
+	var startIdx int
+	for {
```

Comment:
Can you explain why do we need this loop?

Reply:
the loop is here because we have some scenarios where the sidecar should be restarting after an early crash and we're catching the run together at a later stage - this is a way to make the test more resilient (look at the failed logs)

Reply:
the real issue here was that before my PR merged yesterday, the function was looking for "Finishing" instead of "Exiting" which made the check return nil and the test to succeed all the time

Reply:
Can you point me out to the test case for this scenario?

Reply:
https://github.com/kubernetes/kubernetes/blob/95fb4224e6130c9c5a57af817492569034d65e23/test/e2e_node/container_lifecycle_test.go#L2003

Reply:
this is the output before the loop:
```
0) 2024-04-24 13:30:34 +0000 UTC 1333.440000 restartable-init-1 Starting
1) 2024-04-24 13:30:34 +0000 UTC 1333.450000 restartable-init-1 Started
2) 2024-04-24 13:30:34 +0000 UTC 1333.460000 restartable-init-1 Delaying
3) 2024-04-24 13:30:35 +0000 UTC 1334.480000 restartable-init-1 Exiting
4) 2024-04-24 13:30:35 +0000 UTC 1334.620000 init-1 Starting
5) 2024-04-24 13:30:35 +0000 UTC 1334.620000 init-1 Started
6) 2024-04-24 13:30:35 +0000 UTC 1334.630000 init-1 Delaying
7) 2024-04-24 13:30:36 +0000 UTC 1335.630000 restartable-init-1 Starting
8) 2024-04-24 13:30:36 +0000 UTC 1335.640000 init-1 Exiting
9) 2024-04-24 13:30:36 +0000 UTC 1335.650000 restartable-init-1 Started
10) 2024-04-24 13:30:36 +0000 UTC 1335.660000 restartable-init-1 Delaying
11) 2024-04-24 13:30:37 +0000 UTC 1336.610000 init-1 Starting
12) 2024-04-24 13:30:37 +0000 UTC 1336.620000 init-1 Started
13) 2024-04-24 13:30:37 +0000 UTC 1336.620000 init-1 Delaying
14) 2024-04-24 13:30:37 +0000 UTC 1336.660000 restartable-init-1 Exiting
15) 2024-04-24 13:30:38 +0000 UTC 1337.630000 init-1 Exiting
16) 2024-04-24 13:30:54 +0000 UTC 1353.490000 restartable-init-1 Starting
17) 2024-04-24 13:30:54 +0000 UTC 1353.500000 restartable-init-1 Started
18) 2024-04-24 13:30:54 +0000 UTC 1353.520000 restartable-init-1 Delaying
19) 2024-04-24 13:30:54 +0000 UTC 1353.660000 init-1 Started
20) 2024-04-24 13:30:54 +0000 UTC 1353.660000 init-1 Starting
21) 2024-04-24 13:30:54 +0000 UTC 1353.670000 init-1 Delaying
22) 2024-04-24 13:30:55 +0000 UTC 1354.530000 restartable-init-1 Exiting
23) 2024-04-24 13:30:55 +0000 UTC 1354.680000 init-1 Exiting
24) 2024-04-24 13:31:23 +0000 UTC 1382.470000 restartable-init-1 Starting
25) 2024-04-24 13:31:23 +0000 UTC 1382.490000 restartable-init-1 Started
26) 2024-04-24 13:31:23 +0000 UTC 1382.490000 restartable-init-1 Delaying
27) 2024-04-24 13:31:23 +0000 UTC 1382.620000 init-1 Starting
28) 2024-04-24 13:31:23 +0000 UTC 1382.630000 init-1 Started
29) 2024-04-24 13:31:23 +0000 UTC 1382.630000 init-1 Delaying
In [It] at: k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:2003 @ 04/24/24 13:31:23.529
```

Reply:
ok, thanks. That probably makes sense, but I still don't understand how it worked before your fix.

Reply:
it was working because we never checked the proper ordering https://github.com/kubernetes/kubernetes/pull/124086/files#diff-e363a84d59ad2c9d6634456fa26efa9a53d1c7f32b1f1cfe4871ad9c25964e59L130

Reply:
And I also think that looping makes the code less understandable for less complex test cases, when containers don't restart.